R/xgb_binning.R

```diff
@@ -0,0 +1,250 @@
+    params = list(
+      eta = 0.3,
+      max_bin = .max_bin,
+      max_depth = 1,
+      min_child_weight = 5
+    ),
+    nrounds = 100,
+    data = .train,
+    watchlist = list(
+      train = .train,
+      test  = .test
+    ),
+    tree_method = "hist",
+    early_stopping_rounds = 10,
+    objective = "binary:logistic",
+    eval_metric = "auc",
+    verbose = 0
+  )
+}
+
+xgb_binning <- function(df, target, variable){
+
+  levels <- sort(unique(df[[target]]))
+
+  if(length(levels) != 2){
+    stop("Target variable does not have exactly two levels.", call. = FALSE)
+  }
+
+  if(any(is.na((levels)))){
+    stop("Target variable contains missing values.", call. = FALSE)
+  }
+
+  split <- rsample::initial_split(df, prop = 0.8, strata = target)
+  train <- rsample::training(split)
+  test  <- rsample::testing(split)
+
+  xgb_train <- xgboost::xgb.DMatrix(
+    data = as.matrix(train[, variable]),
+    label = ifelse(train[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_test <- xgboost::xgb.DMatrix(
+    data = as.matrix(test[, variable]),
+    label = ifelse(test[[target]] == levels[[1]], 0, 1)
+  )
+
+  xgb_runs <- purrr::map(c(10, 20, 50), function(x) run_xgboost(xgb_train, xgb_test, x))
+
+  xgb_runs_results <- rbind(
+    xgb_runs[[1]]$evaluation_log[xgb_runs[[1]]$best_iteration, ],
+    xgb_runs[[2]]$evaluation_log[xgb_runs[[2]]$best_iteration, ],
+    xgb_runs[[3]]$evaluation_log[xgb_runs[[3]]$best_iteration, ]
+  ) %>%
+  cbind(model = c("mdl_max_bin_10", "mdl_max_bin_20", "mdl_max_bin_50"))
+
+  xgb_best <- xgb_runs_results$model[which.max(xgb_runs_results$test_auc)]
+
+  xgb_selected <- if (xgb_best == "mdl_max_bin_50") {
+    xgb_runs[[3]]
+  } else if (xgb_best == "mdl_max_bin_20") {
+    xgb_runs[[2]]
+  } else {
+    xgb_runs[[1]]
+  }
+
+  xgboost::xgb.model.dt.tree(
+    model = xgb_selected,
+    trees = xgb_selected$best_iteration,
+    use_int_id = TRUE
+  )
+
+}
+
+#' @export
+prep.step_xgb_binning <- function(x,
+                                  training,
+                                  info = NULL,
+                                  ...) {
+
+  col_names <- terms_select(terms = x$terms, info = info)
+  check_type(training[, col_names])
+
+  rules <- vector("list", length(col_names))
+
+  for(i in seq_along(col_names)){
+    if(col_names[[i]] == x$outcome){
+      next
+    } else {
+      rules[[i]] <- xgb_binning(training, x$outcome, col_names[[i]])
+    }
+  }
+
+  names(rules) <- col_names
+
+  step_xgb_binning_new(
+    terms   = x$terms,
+    role    = x$role,
+    trained = TRUE,
+    outcome = x$outcome,
+    rules   = rules,
+    prefix  = x$prefix,
+    skip    = x$skip,
+    id      = x$id
+  )
+}
+
+#' @export
+bake.step_xgb_binning <- function(object,
+                                  new_data,
+                                  ...) {
+
+  vars <- object$rules
+
+  for(i in seq_along(vars)){
+
+    var <- names(vars)[[i]]
+    binned_data <- new_data
+
+    binned_data[, var] <- cut(
```

Comment:
This might make it more complex but think about maybe adding an additional level for missing values.

Reply:
Hmmm, I'm not sure whether this really belongs here. I'd be probably more in favour of leaving it as is and using ```step_* imputation``` methods before that. Specifically I would also like to work on an imputation step that imputes missing values with a constant e.g.: 99999 which is an approach typically used with tree-based methods. By doing so we can ensure that each step focuses on what it should do in isolation. What do you think?

Reply:
> Specifically I would also like to work on an imputation step that imputes missing values with a constant e.g.: 99999 which is an approach typically used with tree-based methods. 

Oh God no. I've been trying to eradicate this approach since the 80's 😀 Please don't ever do this. 

We have a separate step that can do what I proposed so no sweat.

Reply:
> Oh God no. I've been trying to eradicate this approach since the 80's 😀 Please don't ever do this.

Could you elaborate on why you consider this a bad approach?

> We have a separate step that can do what I proposed so no sweat.

What do you have in mind specifically?

Reply:
> Could you elaborate on why you consider this a bad approach?

Special codes, especially ones that try to emulate data of another type, is a poor solution to the problem. 

> What do you have in mind specifically?

Folks can use `step_unknown()` beforehand,